Cubical/Data/Int/Divisibility.agda

```diff
@@ -0,0 +1,308 @@
+
+open import Cubical.Foundations.Prelude
+open import Cubical.Foundations.Equiv
+
+open import Cubical.Data.Nat
+  hiding   (+-assoc ; +-comm ; ·-comm)
+  renaming (_·_ to _·ℕ_; _+_ to _+ℕ_ ; ·-assoc to ·ℕ-assoc)
+open import Cubical.Data.Nat.Order
+open import Cubical.Data.Nat.Mod
+open import Cubical.Data.Int
+
+open import Cubical.Data.Empty as Empty
+open import Cubical.Data.Sum
+open import Cubical.Data.Sigma
+
+open import Cubical.HITs.PropositionalTruncation as Prop
+
+open import Cubical.Relation.Nullary
+
+private
+  variable
+    m n k : ℤ
+
+-- The Divisibility Relation
+-- Most definitions are the same as in Cubical.Data.Nat.Divisibility
+
+_∣_ : ℤ → ℤ → Type
+m ∣ n = ∃[ c ∈ ℤ ] c · m ≡ n
+
+isProp∣ : isProp (m ∣ n)
+isProp∣ = squash
+
+-- Untruncated divisiblility relation
+
+_∣'_ : ℤ → ℤ → Type
+pos 0  ∣' n = 0 ≡ n
+pos (suc m) ∣' n = Σ[ c ∈ ℤ ] c · (pos (suc m)) ≡ n
+(negsuc m)  ∣' n = Σ[ c ∈ ℤ ] c · (negsuc m)    ≡ n
+
+isProp∣' : isProp (m ∣' n)
+isProp∣' {m = pos 0} {n = n} = isSetℤ 0 n
+isProp∣' {m = pos (suc m)} {n = n} p q =
+  Σ≡Prop (λ _ → isSetℤ _ _) (·rCancel (pos (suc m)) _ _ (p .snd ∙ sym (q .snd)) (λ r → snotz (injPos r)))
+isProp∣' {m = negsuc m} {n = n} p q =
+  Σ≡Prop (λ _ → isSetℤ _ _) (·rCancel (negsuc m) _ _ (p .snd ∙ sym (q .snd)) (negsucNotpos _ 0))
+
+∣→∣' : (m n : ℤ) → m ∣ n → m ∣' n
+∣→∣' (pos 0) n ∣ c , p ∣ = ·Comm 0 c ∙ p
+∣→∣' (pos (suc m)) n ∣ p ∣ = p
+∣→∣' (negsuc m) n ∣ p ∣ = p
+∣→∣' m n (squash p q i) = isProp∣' (∣→∣' _ _ p) (∣→∣' _ _ q) i
+
+∣'→∣ : (m n : ℤ) → m ∣' n → m ∣ n
+∣'→∣ (pos 0) n p = ∣ 0 , p ∣
+∣'→∣ (pos (suc m)) n p = ∣ p ∣
+∣'→∣ (negsuc m) n p = ∣ p ∣
+
+∣≃∣' : (m n : ℤ) → (m ∣ n) ≃ (m ∣' n)
+∣≃∣' m n = propBiimpl→Equiv isProp∣ isProp∣' (∣→∣' _ _) (∣'→∣ _ _)
+
+-- Properties of divisibility
+
+∣-left : m ∣ (m · k)
+∣-left {k = k} = ∣ k , ·Comm k _ ∣
+
+∣-right : m ∣ (k · m)
+∣-right {k = k} =  ∣ k , refl ∣
+
+∣-refl : m ≡ n → m ∣ n
+∣-refl p = ∣ 1 , p ∣
+
+∣-zeroˡ : 0 ∣ m → 0 ≡ m
+∣-zeroˡ = ∣→∣' _ _
+
+∣-zeroʳ : m ∣ 0
+∣-zeroʳ = ∣ 0 , refl ∣
+
+∣-+ : k ∣ m → k ∣ n → k ∣ (m + n)
+∣-+ =
+  Prop.map2
+    λ {(c₁ , p) (c₂ , q) → (c₁ + c₂ , ·DistL+ c₁ c₂ _ ∙ (λ t → p t + q t))}
+
+∣-trans : k ∣ m → m ∣ n → k ∣ n
+∣-trans =
+  Prop.map2
+    λ {(c₁ , p) (c₂ , q) → (c₂ · c₁ , sym (·Assoc c₂ c₁ _) ∙ cong (c₂ ·_) p ∙ q)}
+
+∣-right· : k ∣ m → k ∣ (n · m)
+∣-right· {k = k} {n = n} p = ∣-trans p (∣-right {k = n})
+
+-- Exact division
+
+divide : m ∣ n → ℤ
+divide {m = pos 0} _ = 0
+divide {m = pos (suc m)} p = ∣→∣' _ _ p .fst
+divide {m = negsuc m} p = ∣→∣' _ _ p .fst
+
+divideEq : (p : m ∣ n) → divide p · m ≡ n
+divideEq {m = pos 0} = ∣→∣' _ _
+divideEq {m = pos (suc m)} p = ∣→∣' _ _ p .snd
+divideEq {m = negsuc m} p = ∣→∣' _ _ p .snd
+
+-- Bézout and Euclidean Domain
+
+record Bézout (m n : ℤ) : Type where
+  constructor bezout
+  field
+    coef₁ : ℤ
+    coef₂ : ℤ
+    gcd   : ℤ
+    identity : coef₁ · m + coef₂ · n ≡ gcd
+    isGCD : (gcd ∣ m) × (gcd ∣ n)
```

Comment:
Don't you need to say that `gcd` is the biggest such divisor as well?

Reply:
That's equivalent to `identity` + `isGCD` and I think this form is simpler and useful. But It seems the name `isGCD` not exactly precise. Should I include the definition of GCD and add a proof of equivalence?

Reply:
I add an equivalence proof.

Reply:
Right, maybe calling it `isDivisor` would be better?